File: epi_judge_python/test_framework/test_complexity.py
import math
import time

# BigO is passed to a benchmark in order to specify the asymptotic computational complexity for the
# benchmark. In case AUTO is selected, complexity will be calculated automatically to the best fit.
(ONE, LOGN, N, N_LOGN, N_SQUARED, N_SQUARED_LOGN, N_CUBED, TWO_POWER_N,
 N_FACTORIAL, NONE, AUTO) = range(11)


# This data structure will contain the result returned by MinimalLeastSq.
#   - coef        : Estimated coefficient for the high-order term as interpolated from data
#   - rms         : Normalized Root Mean Squared Error
#   - complexity  : Scalability form (e.g. BigO.N, BigO.N_LOGN). In case a scalability form has
#                   been provided to MinimalLeastSq this will return the same value. In case
#                   BigO.AUTO has been selected, this parameter will return the best fitting curve
#                   detected.
class LeastSq:
    def __init__(self):
        self.coef = 0.0
        self.rms = 0.0
        self.complexity = NONE


# Internal function to calculate the different scalability forms.
def fitting_curve(complexity):
    return {
        N: lambda n: n,
        N_SQUARED: lambda n: n**2,
        N_CUBED: lambda n: n**3,
        LOGN: lambda n: float('-inf') if n == 0 else math.log2(n),
        N_LOGN: lambda n: float('-inf') if n == 0 else n * math.log2(n),
        N_SQUARED_LOGN: lambda n: float('-inf')
        if n == 0 else n * n * math.log2(n),
        TWO_POWER_N: lambda n: 2**n,
        N_FACTORIAL: lambda n: math.factorial(int(n)),
    }.get(complexity, lambda n: 1.0)


# Function to return an string for the calculated complexity.
def get_big_o_string(complexity):
    return {
        N: 'n',
        N_SQUARED: 'n^2',
        N_CUBED: 'n^3',
        LOGN: 'logn',
        N_LOGN: 'nlogn',
        N_SQUARED_LOGN: 'n^2logn',
        TWO_POWER_N: '2^n',
        N_FACTORIAL: 'n!',
        ONE: '1',
    }.get(complexity, 'f(n)')


def minimal_least_sq_with_curve(n, time, curve):
    sigma_gn_squared = sum(curve(each_n)**2 for each_n in n)
    sigma_time = sum(each_time for each_time in time)
    sigma_time_gn = sum(each_time * curve(each_n)
                        for each_time, each_n in zip(time, n))

    result = LeastSq()

    result.coef = sigma_time_gn / sigma_gn_squared
    rms = sum((each_time - result.coef * curve(each_n))**2
              for each_time, each_n in zip(time, n))
    mean = sigma_time / len(n)
    result.rms = math.sqrt(rms / len(n)) / mean

    return result


def find_maximum_power(max_n):
    if max_n > 250000000.0:
        return N
    elif max_n > 2500000.0:
        return N_LOGN
    elif max_n > 2500.0:
        return N_SQUARED
    elif max_n > 250.0:
        return N_CUBED
    elif max_n > 25.0:
        return TWO_POWER_N
    elif max_n > 11.0:
        return N_FACTORIAL
    return NONE


def minimal_least_sq(n, time, complexity):
    best_fit = LeastSq()

    if complexity == AUTO:
        fit_curves = (ONE, LOGN, N, N_LOGN, N_SQUARED, N_SQUARED_LOGN, N_CUBED,
                      TWO_POWER_N, N_FACTORIAL)
        # Filter fit_curves as we cannot try all possible curves which might overflow for some n.
        fit_curves = fit_curves[:find_maximum_power(max(n))]

        best_fit = minimal_least_sq_with_curve(n, time, fitting_curve(ONE))
        best_fit.complexity = ONE

        for fit in fit_curves:
            current_fit = minimal_least_sq_with_curve(n, time,
                                                      fitting_curve(fit))
            if current_fit.rms < best_fit.rms:
                best_fit = current_fit
                best_fit.complexity = fit
    else:
        best_fit = minimal_least_sq_with_curve(n, time,
                                               fitting_curve(complexity))
        best_fit.complexity = complexity

    return best_fit


def power_to_complexity(power):
    return int(round(2.0 * power))


def get_big_o_string_multiple_params(metric_names, power, oper):
    def create_symbol_with_complexity(symbol, complexity):
        if complexity == ONE:
            return ''
        elif complexity == LOGN:
            return f'log({symbol})'
        elif complexity == N:
            return f'{symbol}'
        elif complexity == N_LOGN:
            return f'{symbol}log({symbol})'
        elif complexity in (N_SQUARED, N_SQUARED_LOGN, N_CUBED):
            power = 0.5 * complexity
            return '{}^{}'.format(symbol, int(
                math.floor(power))) + ('' if math.isclose(
                    power, math.floor(power)) else f'log({symbol})')
        elif complexity == TWO_POWER_N:
            return f'2^({symbol})'
        elif complexity == N_FACTORIAL:
            return f'({symbol})!'
        return ''

    result = create_symbol_with_complexity(metric_names[0],
                                           power_to_complexity(power[0]))
    for s, p, o in zip(metric_names[1:], power[1:], oper):
        term = create_symbol_with_complexity(s, power_to_complexity(p))
        if term:
            result += (o if result else '') + term
    return 'O(' + (result if result else '1') + ')'


def minimal_least_sq_multiple_params(metric_names, metrics, iter_times,
                                     timeout):
    calc_time_start = time.perf_counter()

    def generate_expression(metrics):
        powers, opers = [], []

        def compute_max_n_for_each_param(metrics):
            return [max(param_col) for param_col in zip(*metrics)]

        def generate_expression_helper(max_n_for_params, param_idx, power_buf,
                                       oper_buf):
            if param_idx == len(max_n_for_params) - 1:
                for p in range(find_maximum_power(
                        max_n_for_params[param_idx])):
                    powers.append(power_buf + [0.5 * p])
                    opers.append(oper_buf)
                return
            for p in range(find_maximum_power(max_n_for_params[param_idx])):
                for o in ('+', '*'):
                    generate_expression_helper(max_n_for_params, param_idx + 1,
                                               power_buf + [0.5 * p],
                                               oper_buf + [o])

        generate_expression_helper(compute_max_n_for_each_param(metrics), 0,
                                   [], [])
        return powers, opers

    powers, opers = generate_expression(metrics)

    def reduce_according_oper(metric, power, oper):
        prev = result = fitting_curve(power_to_complexity(power[0]))(metric[0])
        for x, p, op in zip(metric[1:], power[1:], oper):
            curve = fitting_curve(power_to_complexity(p))
            if op == '+':
                prev = curve(x)
                result += curve(x)
            else:
                prev = prev * curve(x)
                result += prev * curve(x) - prev
        return result

    closest_power_oper = None
    least_rms = float('inf')
    for i, (power, oper) in enumerate(zip(powers, opers)):
        if i % 1024 == 0 and timeout != 0:
            calc_time_current = time.perf_counter()
            if calc_time_current - calc_time_start > timeout:
                return 'TIMEOUT'

        n = [reduce_according_oper(metric, power, oper) for metric in metrics]
        if any(not math.isfinite(d) for d in n):
            continue

        result = minimal_least_sq(n, iter_times, N)
        if result.rms < least_rms:
            least_rms = result.rms
            closest_power_oper = (power, oper)
    return get_big_o_string_multiple_params(metric_names, *closest_power_oper)

File: epi_judge_java_solutions/epi/test_framework/TestComplexity.java

package epi.test_framework;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.function.ToDoubleFunction;

public class TestComplexity {
  // BigO is passed to a benchmark in order to specify the asymptotic
  // computational complexity for the benchmark. In case AUTO is selected,
  // complexity will be calculated automatically to the best fit.
  public enum BigO {
    ONE,
    LOGN,
    N,
    N_LOGN,
    N_SQUARED,
    N_SQUARED_LOGN,
    N_CUBED,
    TWO_POWER_N,
    N_FACTORIAL,
    NONE,
    AUTO
  }

  // This data structure will contain the result returned by MinimalLeastSq.
  //   - coef        : Estimated coefficient for the high-order term as
  //                   interpolated from data
  //   - rms         : Normalized Root Mean Squared Error
  //   - complexity  : Scalability form (e.g. BigO.N, BigO.N_LOGN). In case a
  //                   scalability form has been provided to MinimalLeastSq
  //                   this will return the same value. In case BigO.AUTO
  //                   has been selected, this parameter will return the best
  //                   fitting curve detected.
  public static class LeastSq {
    public double coef = 0.0;
    public double rms = 0.0;
    BigO complexity = BigO.NONE;
  }

  private static double factorial(int n) {
    double result = 1.0;
    for (int i = 2; i <= n; ++i) {
      result *= i;
    }
    return result;
  }

  // Internal function to calculate the different scalability forms.
  private static ToDoubleFunction<Double> fittingCurve(BigO complexity) {
    switch (complexity) {
    case N:
      return (n) -> n;
    case N_SQUARED:
      return (n) -> n * n;
    case N_CUBED:
      return (n) -> n * n * n;
    case LOGN:
      return (n) -> Math.log(n) / Math.log(2);
    case N_LOGN:
      return (n) -> n * Math.log(n) / Math.log(2);
    case N_SQUARED_LOGN:
      return (n) -> n * n * Math.log(n) / Math.log(2);
    case TWO_POWER_N:
      return (n) -> Math.pow(2, n);
    case N_FACTORIAL:
      return (n) -> factorial(n.intValue());
    case ONE:
    default:
      return (n) -> 1.0;
    }
  }

  // Function to return an string for the calculated complexity.
  private static String getBigOString(BigO complexity) {
    switch (complexity) {
    case N:
      return "n";
    case N_SQUARED:
      return "n^2";
    case N_CUBED:
      return "n^3";
    case LOGN:
      return "logn";
    case N_LOGN:
      return "nlogn";
    case N_SQUARED_LOGN:
      return "n^2logn";
    case TWO_POWER_N:
      return "2^n";
    case N_FACTORIAL:
      return "n!";
    case ONE:
      return "1";
    default:
      return "f(n)";
    }
  }

  // Find the coefficient for the high-order term in the running time, by
  // minimizing the sum of squares of relative error, for the fitting curve
  // given by the lambda expresion.
  //   - n             : List containing the size of the benchmark tests
  //   - time          : List containing the times for the benchmark tests
  //   - curve         : lambda expresion of a curve
  private static LeastSq
  minimalLeastSqWithCurve(List<Double> n, List<Long> time,
                          ToDoubleFunction<Double> curve) {
    double sigmaGn = 0.0;
    double sigmaGnSquared = 0.0;
    double sigmaTime = 0.0;
    double sigmaTimeGn = 0.0;

    // Calculate least square fitting parameter.
    for (int i = 0; i < n.size(); ++i) {
      double gnI = curve.applyAsDouble(n.get(i));
      sigmaGn += gnI;
      sigmaGnSquared += gnI * gnI;
      sigmaTime += time.get(i);
      sigmaTimeGn += time.get(i) * gnI;
    }

    LeastSq result = new LeastSq();

    // Calculate complexity.
    result.coef = sigmaTimeGn / sigmaGnSquared;

    // Calculate RMS.
    double rms = 0.0;
    for (int i = 0; i < n.size(); ++i) {
      double fit = result.coef * curve.applyAsDouble(n.get(i));
      rms += Math.pow(time.get(i) - fit, 2);
    }

    // Normalized RMS by the mean of the observed values.
    double mean = sigmaTime / n.size();
    result.rms = Math.sqrt(rms / n.size()) / mean;

    return result;
  }

  private static int findMaximumPower(Double maxN) {
    if (maxN > 250000000.0) {
      return BigO.N.ordinal();
    } else if (maxN > 2500000.0) {
      return BigO.N_LOGN.ordinal();
    } else if (maxN > 2500.0) {
      return BigO.N_SQUARED.ordinal();
    } else if (maxN > 250.0) {
      return BigO.N_CUBED.ordinal();
    } else if (maxN > 25.0) {
      return BigO.TWO_POWER_N.ordinal();
    } else if (maxN > 11.0) {
      return BigO.N_FACTORIAL.ordinal();
    }
    return BigO.NONE.ordinal();
  }

  // Find the coefficient for the high-order term in the running time, by
  // minimizing the sum of squares of relative error.
  //   - n          : List containing the size of the benchmark tests
  //   - time       : List containing the times for the benchmark tests
  //   - complexity : If different than AUTO, the fitting curve will stick to
  //                  this one. If it is AUTO, it will be calculated the best
  //                  fitting curve.
  private static LeastSq minimalLeastSq(List<Double> n, List<Long> time,
                                        BigO complexity) {
    LeastSq bestFit;

    if (complexity == BigO.AUTO) {
      List<BigO> fitCurves =
          List.of(BigO.ONE, BigO.LOGN, BigO.N, BigO.N_LOGN, BigO.N_SQUARED,
                  BigO.N_SQUARED_LOGN, BigO.N_CUBED, BigO.TWO_POWER_N,
                  BigO.N_FACTORIAL);
      // Filter fitCurves as we cannot try all possible curves which might
      // overflow for some n.
      fitCurves = fitCurves.subList(0, findMaximumPower(Collections.max(n)));

      // Take BigO.ONE as default best fitting curve.
      bestFit = minimalLeastSqWithCurve(n, time, fittingCurve(BigO.ONE));
      bestFit.complexity = BigO.ONE;

      // Compute all possible fitting curves and stick to the best one.
      for (BigO fit : fitCurves) {
        LeastSq currentFit =
            minimalLeastSqWithCurve(n, time, fittingCurve(fit));
        if (currentFit.rms < bestFit.rms) {
          bestFit = currentFit;
          bestFit.complexity = fit;
        }
      }
    } else {
      bestFit = minimalLeastSqWithCurve(n, time, fittingCurve(complexity));
      bestFit.complexity = complexity;
    }

    return bestFit;
  }

  private static void
  generateExpression(List<Integer> maxNForParams, Integer paramIdx,
                     List<Double> powerBuf, List<List<Double>> powers,
                     List<String> operBuf, List<List<String>> opers) {
    if (paramIdx == maxNForParams.size() - 1) {
      for (int p = 0;
           p < findMaximumPower(maxNForParams.get(paramIdx).doubleValue());
           ++p) {
        powerBuf.add(0.5 * p);
        powers.add(new ArrayList<>(powerBuf));
        powerBuf.remove(powerBuf.size() - 1);
        opers.add(new ArrayList<>(operBuf));
      }
      return;
    }
    for (int p = 0;
         p < findMaximumPower(maxNForParams.get(paramIdx).doubleValue()); ++p) {
      powerBuf.add(0.5 * p);
      for (String o : List.of("+", "*")) {
        operBuf.add(o);
        generateExpression(maxNForParams, paramIdx + 1, powerBuf, powers,
                           operBuf, opers);
        operBuf.remove(operBuf.size() - 1);
      }
      powerBuf.remove(powerBuf.size() - 1);
    }
  }

  private static BigO powerToComplexity(Double power) {
    return BigO.values()[(int)(2.0 * power)];
  }

  private static Double reduceAccordingOper(List<Integer> metric,
                                            List<Double> power,
                                            List<String> oper) {
    Double prev = fittingCurve(powerToComplexity(power.get(0)))
                      .applyAsDouble(metric.get(0).doubleValue());
    Double result = prev;
    for (int i = 1; i < metric.size(); ++i) {
      ToDoubleFunction<Double> curve =
          fittingCurve(powerToComplexity(power.get(i)));
      if (oper.get(i - 1).equals("+")) {
        prev = curve.applyAsDouble(metric.get(i).doubleValue());
        result += curve.applyAsDouble(metric.get(i).doubleValue());
      } else {
        prev = prev * curve.applyAsDouble(metric.get(i).doubleValue());
        result +=
            prev * curve.applyAsDouble(metric.get(i).doubleValue()) - prev;
      }
    }
    return result;
  }

  private static String createSymbolWithComplexity(String symbol,
                                                   BigO complexity) {
    switch (complexity) {
    case ONE:
      return "";
    case LOGN:
      return String.format("log(%s)", symbol);
    case N:
      return symbol;
    case N_LOGN:
      return String.format("%slog(%s)", symbol, symbol);
    case N_SQUARED:
    case N_SQUARED_LOGN:
    case N_CUBED:
      final double power = 0.5 * complexity.ordinal();
      return String.format("%s^%d", symbol, (int)Math.floor(power)) +
          (Double.compare(power, Math.floor(power)) == 0
               ? ""
               : String.format("log(%s)", symbol));
    case TWO_POWER_N:
      return String.format("2^(%s)", symbol);
    case N_FACTORIAL:
      return String.format("(%s)!", symbol);
    default:
      return "";
    }
  }

  private static String getBigOStringMultipleParams(List<String> metricNames,
                                                    List<Double> power,
                                                    List<String> oper) {
    StringBuilder result = new StringBuilder(createSymbolWithComplexity(
        metricNames.get(0), powerToComplexity(power.get(0))));
    for (int i = 1; i < power.size(); ++i) {
      final String term = createSymbolWithComplexity(
          metricNames.get(i), powerToComplexity(power.get(i)));
      if (!term.isEmpty()) {
        result.append(result.toString().isEmpty() ? "" : oper.get(i - 1))
            .append(term);
      }
    }
    return "O(" + (result.toString().isEmpty() ? "1" : result.toString()) + ")";
  }

  private static List<Integer>
  computeMaxNForEachParam(List<List<Integer>> metrics) {
    List<Integer> maxNForParams =
        new ArrayList<>(Collections.nCopies(metrics.get(0).size(), 1));
    for (List<Integer> metric : metrics) {
      for (int i = 0; i < metric.size(); ++i) {
        maxNForParams.set(i, Math.max(maxNForParams.get(i), metric.get(i)));
      }
    }
    return maxNForParams;
  }

  public static String minimalLeastSqMultipleParams(List<String> metricNames,
                                                    List<List<Integer>> metrics,
                                                    List<Long> iterTimes,
                                                    long timeout) {
    long calcTimeStart = System.nanoTime();
    List<List<Double>> powers = new ArrayList<>();
    List<List<String>> opers = new ArrayList<>();
    generateExpression(computeMaxNForEachParam(metrics), 0, new ArrayList<>(),
                       powers, new ArrayList<>(), opers);

    Double leastRms = Double.MAX_VALUE;
    List<Double> closestPower = null;
    List<String> closestOper = null;
    for (int i = 0; i < powers.size(); ++i) {
      if (i % 1024 == 0 && timeout != 0) {
        long calcTimeCurrent = System.nanoTime();
        if (calcTimeCurrent - calcTimeStart > timeout * 1E9) {
          return "TIMEOUT";
        }
      }
      List<Double> n = new ArrayList<>();
      for (List<Integer> metric : metrics) {
        n.add(reduceAccordingOper(metric, powers.get(i), opers.get(i)));
      }
      if (n.stream().anyMatch(d -> d.isInfinite() || d.isNaN())) {
        continue;
      }

      LeastSq result = minimalLeastSq(n, iterTimes, BigO.N);
      if (result.rms < leastRms) {
        leastRms = result.rms;
        closestPower = powers.get(i);
        closestOper = opers.get(i);
      }
    }
    return getBigOStringMultipleParams(metricNames, closestPower, closestOper);
  }
}
